### DISCUSSION TITLE
AttributeError: 'AsyncSessionTransaction' object has no attribute 'run_sync'

### DISCUSSION THREAD

**Discussion by cnicodeme** [sqlalchemy/sqlalchemy#10526] [Usage Questions]
Hi!

I'm trying to run the create/drop commands for my database. It was previously working but I suspect that an upgrade to SQLAlchemy made some breaking changes to how I was accessing my code.

Here's the current integration:

```python
        async with dbctx.begin() as conn:
            await conn.run_sync(ORModel.metadata.drop_all)
            await conn.run_sync(ORModel.metadata.create_all)
```

`dbctx` is an initiated session from 

```python
session = sessionmaker(
            bind=self.write_engine,
            expire_on_commit=False,
            class_=AsyncSession,
            autoflush=True
        )

dbctx = session()
```

I'm using `mysql+asyncmy` for the engine.

I also tried this variant:

```python
            await dbctx.run_sync(ORModel.metadata.drop_all)
            await dbctx.run_sync(ORModel.metadata.create_all)
```

Without any luck.

What am I doing wrong?

Thanks for your help!

**Comment by CaselIT** [comment]
Hi,

Session.begin returns the transaction that has never included run_sync or any other execute function.

You were probably doing `session.begin()` where session is your sessionmaker.

Also you should use the asyncsessionmaker with an async session instead of the normal one

**Reply by cnicodeme** [reply]
Thank you for your help!

If I replace the `dbctx` by `with session.begin() as conn`, I get the following error:

> AttributeError: 'Session' object has no attribute '_run_ddl_visitor'

Do you know why?

**Reply by cnicodeme** [reply]
If I remove the `begin` and use directly the `dbctx` session instance, I get the same `_run_ddl_visitor` error:

```python
await dbctx.run_sync(ORModel.metadata.drop_all)
await dbctx.run_sync(ORModel.metadata.create_all)
```

(I changed my `sessionmaker` to the `async_sessionmaker`, it hasn't changed the current issue)

**Reply by CaselIT** [reply]
you are trying to use a session where a connection is needed. use `connection()` to get a connection from a session. That can be used to the metadata callbacks

**Reply by cnicodeme** [reply]
Yes! Thank you! That was it!

Here's the working code:

```python
session = sessionmaker(
            bind=self.write_engine,
            expire_on_commit=False,
            class_=AsyncSession,
            autoflush=True
        )

dbctx = session()
conn = dbctx.connection()
await dbctx.run_sync(ORModel.metadata.drop_all)
await dbctx.run_sync(ORModel.metadata.create_all)
```

Thank you so much for your (fast!) help

**Reply by CaselIT** [reply]
@zzzeek would it make sense to make the session fail in a nicer way for this?

**Reply by zzzeek** [reply]
dont really know.  it all looks like "hey lets just add methods so it 'just works'" to me, which is a terrible idea because that makes it even more confusing when somethign doesnt work

**Reply by zzzeek** [reply]
for the ddl_visitor error we can have create/create_all/etc do a check for Connection/Engine

**Reply by CaselIT** [reply]
> Here's the working code:

Interesting, I would have expected that this was what was required:
```py
sm = async_sessionmaker(...)
async with sm() as session:
  conn = await session.connection()
  await conn.run_sync(ORModel.metadata.drop_all)
  await conn.run_sync(ORModel.metadata.create_all)
```

**Reply by CaselIT** [reply]
> dont really know. it all looks like "hey lets just add methods so it 'just works'" to me, which is a terrible idea because that makes it even more confusing when somethign doesnt work

yes my idea was to probably catch the attribute error and provide a bit more information

**Reply by CaselIT** [reply]
looking at the core it's likely not worth it
